CORE-10223: Add panda link component

michael-redpanda opened the pull request:
<!--
See https://github.com/redpanda-data/redpanda/blob/dev/CONTRIBUTING.md#pull-request-body
for more details and examples of what is expected in a PR body.

Content in this top section is REQUIRED. Describe, in plain language, the motivation
behind the change (bug fix, feature, improvement) in this PR and how the included
commits address it.

Add the GitHub keyword `Fixes` to link to bug(s) this PR will fix, e.g.
  Fixes #ISSUE-NUMBER, Fixes #ISSUE-NUMBER, ...

If this PR is a backport, link to the original with `Backport of PR`, e.g.
  Backport of PR #PR-NUMBER
-->

This PR adds the `panda_link` component to Redpanda.  This component is made up of:

* `link`s - these are instances that represent a connection from the destination cluster to the source cluster
* `manager` - The link manager, instantiated by the `service` which manages links
* `service` - The API layer which holds an instance of `manager` and registers for callbacks and forwards them to the contained `manager`.

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [X] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v25.1.x
- [ ] v24.3.x
- [ ] v24.2.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

* None

michael-redpanda committed:
panda_link: Added logger

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
panda_link: Introduce manager class

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/pl: Added callbacks to table

Implemented callbacks when changes are made and added ability to
register callbacks via the front end

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
cluster/pl: Added link accessors to frontend

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
pl: Added link class

Added a link class that represents a cluster link established between
the destination cluster to the source cluster.  A new link_factory is
created that is used by the manager to create links.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda committed:
pl: Added service

The service class is the API surface for cluster linking.  It has
registered for callbacks from different cluster services and forwards
them to the manager.

Signed-off-by: Michael Boquard <michael@redpanda.com>

mmaslankaprv commented on src/v/panda_link/logger.cc:
```diff
@@ -0,0 +1,16 @@
+#include "panda_link/logger.h"
+
+namespace panda_link {
+ss::logger pllog("panda_link");
```
nit: are we going to change the that name to `cluster_link` ?

michael-redpanda replied:
yes there's a follow up commit that does a find/replace

mmaslankaprv commented on src/v/cluster/panda_link/table.cc:
```diff
@@ -32,7 +32,23 @@ size_t table::size() const { return _link_metadata.size(); }
     name_index_t snap_name_index;
+    ss::chunked_fifo<id_t> all_deletes;
+    ss::chunked_fifo<id_t> all_inserts;
+    ss::chunked_fifo<id_t> all_changes;
```
consider using `chunked_vector` instead. Chunked fifo has pretty large default chunk size

mmaslankaprv commented on src/v/panda_link/service.cc:
```diff
@@ -0,0 +1,170 @@
+      std::make_unique<link_registry_adapter>(&_plf->local()),
+      std::make_unique<default_link_factory>());
+
+    co_await _manager->start();
+    register_notifications();
```
Are we guaranteed that manger will not loose any notification ? it is being queried in start and only then we `register` notifications

michael-redpanda replied:
For cluster link table changes, I think we're safe.

For leadership changes... yeah good point.  It may need to do a follow up query of leadership information after the callbacks have been registered.

mmaslankaprv replied:
there are scheduling points between manager start and `register_notifications` there may be some notification firing after the `manager` is started but before the `register_notification` is called. If that is not a case this place would benefit from comment.

michael-redpanda replied:
> For cluster link table changes, I think we're safe.

I should have expanded on what I meant here - table changes should only be happening via user input, probably through Admin API which starts up much later after manager starts, which is why I feel we are safe irt not missing table changes.

For leadership though, I believe you are correct and it needs to be handled.

mmaslankaprv replied:
but when Redpanda starts the changes are applied from the log i.e. the log is read and then applied, if one of the controller partition replicas is behind it may be recovered and controller commands would be applied during that time. there is not synchronisation between the recovery and start sequence

michael-redpanda replied:
ah gotcha, ok i'll have a think about that

mmaslankaprv commented on src/v/panda_link/service.cc:
```diff
@@ -0,0 +1,170 @@
+    if (!node_is_leader) {
+        _manager->on_leadership_change(partition->ntp(), ntp_leader::no);
+    }
+    auto is_leader = partition->is_elected_leader() ? ntp_leader::yes
```
this should be `is_leader()` `is_elected_leader` returns true right after election but before the term is confirmed

mmaslankaprv commented on src/v/panda_link/manager.cc:
```diff
@@ -0,0 +1,93 @@
+    vlog(pllog.trace, "NTP={} leadership changed to {}", ntp, is_ntp_leader);
+}
+
+ss::future<> manager::handle_on_link_change(model::id_t id) {
```
what happens if this function throws ?

michael-redpanda replied:
I don't think there's any spot in here that will throw... if it does it'll be caught by the work_queue and the "unexpected panda link manager error" message will get printed.

dotnwat replied:
i think the point is that this function has a lot of dependencies (it->second->stop, new_link->start, create_link, etc...) which might throw in the future (either on purpose or on accident), and this function may not be exception safe if it's updating two or more data structures that are internally cross-referenced.

mmaslankaprv commented on src/v/panda_link/manager.h:
```diff
@@ -0,0 +1,102 @@
+    ::model::node_id _self;
+    std::unique_ptr<link_registry> _registry;
+    std::unique_ptr<link_factory> _link_factory;
+    ntp_leader _is_controller_leader{ntp_leader::no};
```
this seems to be unused

michael-redpanda replied:
Ah yeah, I guess I could just remove it and add it when it is being used

mmaslankaprv replied:
are you going to remove this ?

michael-redpanda replied:
I believe I did remove `_is_controller_leader`

mmaslankaprv commented on src/v/panda_link/manager.h:
```diff
@@ -0,0 +1,102 @@
+ * @brief Abstract class that provides accessors to panda link table
+ *
+ */
+class link_registry {
```
I am not sure if we need that interface, in the tests it isn't mocked at all as we are just using the table, this way an adapter is defined twice.

michael-redpanda replied:
In `service` it uses the front end, in the tests it uses the table

michael-redpanda commented:
Force push:

* Addressed PR comments

vbotbuildovich commented:
<!--TEST RESULTS-->
<!--5cd19743-0c68-4d4e-b3b7-a8724f397607-->
<!--TEST RESULTS-->
<!--e6205057-b5d1-493c-b3a8-86f8c8d1ea4e-->
<!--TEST RESULTS-->
<!--fc143c87-e3e3-4aa4-8c50-caee90ef082e-->
<!--TEST RESULTS-->
<!--c557d744-ce80-43a2-97d4-7a0044b6953d-->
<!--title-->
## CI test results
<!--/title-->
<!--body-->
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67387">build#67387</a></summary>


| test_class                | test_method                   | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:--------------------------|:------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| CloudStorageChunkReadTest | test_read_chunks              |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67387#0197798e-861b-4eb2-99f1-ed57cc50774a | FLAKY         | 18/21    | upstream reliability is '97.21362229102168'. current run reliability is '85.71428571428571'. drift is 11.49934 and the allowed drift is set to 50. The test should PASS |
| CrashReporterTest         | test_redpanda_crash_reporting |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67387#0197798e-861c-44c4-bc32-59e9f2f33c43 | FLAKY         | 20/21    | upstream reliability is '99.62121212121212'. current run reliability is '95.23809523809523'. drift is 4.38312 and the allowed drift is set to 50. The test should PASS  |
| RandomNodeOperationsTest  | test_node_operations          | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": false, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67387#01977978-b04a-4efe-842e-7f9962313c19 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67520">build#67520</a></summary>


| test_class                        | test_method                               | test_arguments                                                                                                                                 | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:----------------------------------|:------------------------------------------|:-----------------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| cluster_metadata_uploader_fixture | test_upload_in_term                       |                                                                                                                                                | unit        | https://buildkite.com/redpanda/redpanda/builds/67520#01978089-59ca-4075-99c0-f28913c70f0e | FAIL          | 0/1      |                                                                                                                                                                         |
| ArchivalTest                      | test_timeboxed_uploads                    | {"acks": 1, "cloud_storage_type": 2}                                                                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67520#019780cc-8205-40a7-a5de-f2268c61830b | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| DatalakeDelayedEnablementTest     | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67520#019780cc-8204-4062-a6d2-ddcb968e52ce | FLAKY         | 17/21    | upstream reliability is '91.07142857142857'. current run reliability is '80.95238095238095'. drift is 10.11905 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest          | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": false}        | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67520#019780cc-8205-40a7-a5de-f2268c61830b | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
| RandomNodeOperationsTest          | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "chunked_sliding_window", "enable_failures": true, "mixed_versions": false, "with_iceberg": true} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67520#019780cc-8203-45d1-af1a-77f09addfc15 | FLAKY         | 19/21    | upstream reliability is '100.0'. current run reliability is '90.47619047619048'. drift is 9.52381 and the allowed drift is set to 50. The test should PASS              |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67548">build#67548</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                          | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                   |
|:------------------------------|:------------------------------------------|:----------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:-------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                  | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67548#01978331-d54a-4a09-84a5-086aa070c704 | FLAKY         | 13/21    | upstream reliability is '83.97790055248619'. current run reliability is '61.904761904761905'. drift is 22.07314 and the allowed drift is set to 50. The test should PASS |
| DatalakeDiskUsageTest         | test_idle_finish                          | {"cloud_storage_type": 1, "concurrent_translations": 4, "num_partitions": 10}                                                           | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67548#01978331-d54c-4779-bdf1-8001810ceed2 | FLAKY         | 20/21    | upstream reliability is '99.40828402366864'. current run reliability is '95.23809523809523'. drift is 4.17019 and the allowed drift is set to 50. The test should PASS   |
| PartitionBalancerTest         | test_fuzz_admin_ops                       |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67548#01978331-d54b-4e2b-8c08-6855a7ab589f | FLAKY         | 19/21    | upstream reliability is '96.05263157894737'. current run reliability is '90.47619047619048'. drift is 5.57644 and the allowed drift is set to 50. The test should PASS   |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 2, "compaction_mode": "adjacent_merge", "enable_failures": true, "mixed_versions": false, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67548#01978331-d54c-4779-bdf1-8001810ceed2 | FLAKY         | 18/21    | upstream reliability is '98.36065573770492'. current run reliability is '85.71428571428571'. drift is 12.64637 and the allowed drift is set to 50. The test should PASS  |
| TimeQueryKafkaTest            | test_timequery_below_start_offset         |                                                                                                                                         | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67548#01978331-d54b-4e2b-8c08-6855a7ab589f | FLAKY         | 20/21    | upstream reliability is '98.91696750902527'. current run reliability is '95.23809523809523'. drift is 3.67887 and the allowed drift is set to 50. The test should PASS   |
</details>
<details>
<summary>test results on <a href="https://buildkite.com/redpanda/redpanda/builds/67651">build#67651</a></summary>


| test_class                    | test_method                               | test_arguments                                                                                                                         | test_kind   | job_url                                                                                   | test_status   | passed   | reason                                                                                                                                                                  |
|:------------------------------|:------------------------------------------|:---------------------------------------------------------------------------------------------------------------------------------------|:------------|:------------------------------------------------------------------------------------------|:--------------|:---------|:------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| ClusterMetadataUploadTest     | test_uploads_after_wipe                   | {"cloud_storage_type": 2}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67651#01978d5d-54a2-49c8-8fde-2d574f382768 | FLAKY         | 20/21    | upstream reliability is '99.10714285714286'. current run reliability is '95.23809523809523'. drift is 3.86905 and the allowed drift is set to 50. The test should PASS  |
| DataMigrationsApiTest         | test_migrated_topic_data_integrity        | {"params": {"cancellation": null, "use_alias": true}, "transfer_leadership": true}                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67651#01978d71-de99-405f-ac0e-91d45cc53241 | FLAKY         | 20/21    | upstream reliability is '100.0'. current run reliability is '95.23809523809523'. drift is 4.7619 and the allowed drift is set to 50. The test should PASS               |
| DatalakeDelayedEnablementTest | test_enabling_iceberg_in_existing_cluster | {"catalog_type": "rest_jdbc", "cloud_storage_type": 1}                                                                                 | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67651#01978d5d-54a4-4471-92ab-10a39e9bedcd | FLAKY         | 12/21    | upstream reliability is '75.90579710144928'. current run reliability is '57.14285714285714'. drift is 18.76294 and the allowed drift is set to 50. The test should PASS |
| RandomNodeOperationsTest      | test_node_operations                      | {"cloud_storage_type": 1, "compaction_mode": "sliding_window", "enable_failures": true, "mixed_versions": true, "with_iceberg": false} | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67651#01978d5d-54a3-44e2-ad49-4cf94f4fd581 | FAIL          | 0/1      |                                                                                                                                                                         |
| TestTieredStoragePause        | test_safe_pause_resume                    | {"allow_gaps_cluster_level": true, "allow_gaps_topic_level": null}                                                                     | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67651#01978d5d-54a3-44e2-ad49-4cf94f4fd581 | FAIL          | 0/1      |                                                                                                                                                                         |
| TopicDeleteCloudStorageTest   | drop_lifecycle_marker_test                | {"cloud_storage_type": 1}                                                                                                              | ducktape    | https://buildkite.com/redpanda/redpanda/builds/67651#01978d71-de99-405f-ac0e-91d45cc53241 | FLAKY         | 18/21    | upstream reliability is '97.38095238095238'. current run reliability is '85.71428571428571'. drift is 11.66667 and the allowed drift is set to 50. The test should PASS |
</details>
<!--/body-->

michael-redpanda committed:
ssx: Add ability to resume/pause work queue

This change permits constructing the work_queue in a paused state and
adds modifiers to resume work.  This can be used to, for example, store
up notifications from callbacks before a service has started and then,
once the service has started, kick off the work queue.

Signed-off-by: Michael Boquard <michael@redpanda.com>

michael-redpanda commented:
Force push (addresses issues with missing callbacks):

* Added ability to pause/resume the work queue
* Moved callback registration to occur before manager start to ensure we don't miss any callbacks while the manager is starting

michael-redpanda commented:
Force push:

* Addressed exception logic in handling link change

mmaslankaprv commented on src/v/ssx/work_queue.h:
```diff
@@ -52,6 +63,7 @@ class work_queue {
     error_reporter_fn _error_reporter;
     ss::condition_variable _cond_var;
     ss::chunked_fifo<ss::noncopyable_function<ss::future<>()>> _tasks;
+    is_running_t _is_running{is_running_t::yes};
```
nit: consider changing this to `paused` ? as running is the default state for the queue

michael-redpanda commented:
Force pushed:

* Fixed PR comments
* Renamed `is_running` to `is_paused`
* Fixed up bad logic on work queue shutdown/abort
* Fixed copilot noted issues

mmaslankaprv commented on src/v/panda_link/manager.cc:
```diff
@@ -16,25 +16,102 @@
+            co_await new_link->start();
+            _links.emplace(id, std::move(new_link));
+        } catch (const std::exception& e) {
+            vlog(pllog.warn, "Failed to create link {}: {}", id, e);
+            vlog(pllog.warn, "Re-attempting link creation within 10 seconds");
```
nit: let's make it a single log line

mmaslankaprv commented on src/v/panda_link/manager.cc:
```diff
@@ -16,25 +16,102 @@
+            } catch (const std::exception& e) {
+                vlog(pllog.warn, "Failed to stop link {}: {}", id, e);
+                // If stopping fails, we can retry after a delay
+                vlog(pllog.warn, "Re-attempting link stop within 10 seconds");
```
nit: try to avoid multi line logs, they may be interleaved with some other log entries from different shards and it will be hard to correlate the two.

mmaslankaprv reviewed (commented):
this looks great, just one comment about the logging

michael-redpanda commented:
Force push:

* Fixed log lines
